File: Medium/Word Break/word-break.py
#User function Template for python3

class Solution:
    def wordBreak(self, n, s, dictionary):
        # Complete this function
        nn=len(s)
        dp=[0]*(nn+1)
        dp[nn]=1
        for i in range(nn-1,-1,-1) :
            for w in dictionary :
                if i+len(w) <= nn and s[i:i+len(w)] == w :
                    dp[i]=dp[i+len(w)]
                if dp[i] :
                    break
        return dp[0]

#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
	test_case = int(input())

	for _ in range(test_case):
		n = int(input())
		dictionary = [word for word in input().strip().split()]
		s = input().strip()
		ob = Solution()
		res = ob.wordBreak(n, s, dictionary)
		if res:
			print(1)
		else:
			print(0)
# } Driver Code Ends

File: Difficulty: Medium/Word Break/word-break.java
//{ Driver Code Starts
import java.io.*;
import java.util.*;

public class GfG {
    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);
        int t = sc.nextInt();
        sc.nextLine();
        while (t-- > 0) {
            String s = sc.nextLine();
            String line = sc.nextLine();
            String[] dictionary = line.split(" ");

            Solution obj = new Solution();
            if (obj.wordBreak(s, dictionary)) {
                System.out.println("true");
            } else {
                System.out.println("false");
            }
            System.out.println("~");
        }
    }
}
// } Driver Code Ends


class Solution {
    public boolean wordBreak(String s, String[] dictionary) {
        // code here
        int n = s.length();
        
        int[] dp = new int[n+1];
        
        dp[0] = 1;
        
        for(int i=1; i<=n; i++){
            for(String str : dictionary){
                int start = i - str.length();
                if(start >= 0 && dp[start] == 1 && s.substring(start, i).equals(str)){
                    dp[i] = 1;
                    break;
                }
            }
        }
        
        if(dp[n]==1)
        {
            return true;
        }
        else
        {
            return false;
        }
    }
}